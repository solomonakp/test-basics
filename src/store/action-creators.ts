import { Dispatch } from 'redux'
import { AxiosResponse } from 'axios'
import { ThunkAction } from 'redux-thunk'

import { Axios } from '../helpers/axios'
import { Product } from '../types/Product'
import {
  FetchedProductsAction,
  FetchedProductAction,
  AddedToCartAction,
} from '../store/actions'
import {
  FETCHING_PRODUCTS,
  FETCHED_PRODUCTS,
  FETCHING_PRODUCT,
  FETCHED_PRODUCT,
  ADDING_TO_CART,
  ADDED_TO_CART,
  FETCHING_CART,
  TOGGLE_CART_OPEN,
  REMOVING_FROM_CART,
} from '../store/constants'

// ThunkAction<(1), (2), (3), (4)>
// (1) The return value of the internal async function that will be returned by this thunk
// (2) The type of data being fired in the last action
// (3) The type of the extra arguments passed to redux
// (4) The last action to be dispatched.
export const fetchProducts = ({
  search = '',
}: {
  search: string
}): ThunkAction<
  Promise<FetchedProductsAction>,
  Product[],
  undefined,
  FetchedProductsAction
> => {
  return (dispatch: Dispatch) => {
    dispatch({
      type: FETCHING_PRODUCTS,
    })

    return Axios.get(`products?search=${search}`).then(
      ({ data }: AxiosResponse<Product[]>) =>
        dispatch({
          type: FETCHED_PRODUCTS,
          products: data,
        }),
    )
  }
}

export const fetchSingleProduct = (
  id: string,
): ThunkAction<
  Promise<FetchedProductAction>,
  Product,
  undefined,
  FetchedProductAction
> => {
  return (dispatch: Dispatch) => {
    dispatch({
      type: FETCHING_PRODUCT,
    })

    return Axios.get(`products/${id}`)
      .then(({ data }: AxiosResponse<Product>) => {
        return dispatch({
          type: FETCHED_PRODUCT,
          product: data,
        })
      })
      .catch((error) => {
        return dispatch({
          type: FETCHED_PRODUCT,
          product: null,
          fetchSingleProductError: true,
        })
      })
  }
}

export const addProductToCart = (
  id: string,
): ThunkAction<
  Promise<AddedToCartAction>,
  Product,
  undefined,
  AddedToCartAction
> => {
  return (dispatch: Dispatch) => {
    dispatch({
      type: ADDING_TO_CART,
    })

    return Axios.post(`cart/${id}`)
      .then(({ data }: AxiosResponse<Product[]>) => {
        return dispatch({
          type: ADDED_TO_CART,
          products: data,
        })
      })
      .catch((error) => {
        return dispatch({
          type: ADDED_TO_CART,
          addToCartError: true,
        })
      })
  }
}

export const removeProductFromCart = (
  id: string,
): ThunkAction<
  Promise<AddedToCartAction>,
  Product,
  undefined,
  AddedToCartAction
> => {
  return (dispatch: Dispatch) => {
    dispatch({
      type: REMOVING_FROM_CART,
    })

    return Axios.delete(`cart/${id}`)
      .then(({ data }: AxiosResponse<Product[]>) => {
        return dispatch({
          type: ADDED_TO_CART,
          products: data,
        })
      })
      .catch((error) => {
        return dispatch({
          type: ADDED_TO_CART,
          addToCartError: true,
        })
      })
  }
}

export const getProductsInCart = (): ThunkAction<
  Promise<AddedToCartAction>,
  Product,
  undefined,
  AddedToCartAction
> => {
  return (dispatch: Dispatch) => {
    return Axios.get(`cart`)
      .then(({ data }: AxiosResponse<Product[]>) => {
        return dispatch({
          type: ADDED_TO_CART,
          products: data,
        })
      })
      .catch((error) => {
        return dispatch({
          type: ADDED_TO_CART,
          addToCartError: true,
        })
      })
  }
}

export const toggleCartOpen = () => ({
  type: TOGGLE_CART_OPEN,
})
