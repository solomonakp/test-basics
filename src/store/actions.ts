import { Action } from 'redux'
import { Product } from '../types/Product'
import {
  FETCHING_PRODUCTS,
  FETCHED_PRODUCTS,
  FETCHING_PRODUCT,
  FETCHED_PRODUCT,
  ADDING_TO_CART,
  ADDED_TO_CART,
  TOGGLE_CART_OPEN,
} from './constants'

export interface FetchProductAction extends Action<typeof FETCHING_PRODUCTS> {}

export interface FetchedProductsAction extends Action<typeof FETCHED_PRODUCTS> {
  products: Product[]
}

export interface FetchedProductAction extends Action<typeof FETCHED_PRODUCT> {
  product?: Product
  fetchSingleProductError?: Boolean
}

export interface FetchingProductAction
  extends Action<typeof FETCHING_PRODUCT> {}

export interface AddingToCartAction extends Action<typeof ADDING_TO_CART> {}

export interface AddedToCartAction extends Action<typeof ADDED_TO_CART> {
  products?: Product[]
  addToCartError?: boolean
}

export interface ToggleCartOpen extends Action<typeof TOGGLE_CART_OPEN> {}

export type ProductActions =
  | FetchProductAction
  | FetchedProductsAction
  | FetchingProductAction
  | FetchedProductAction

export type CartActions =
  | AddingToCartAction
  | AddedToCartAction
  | ToggleCartOpen
