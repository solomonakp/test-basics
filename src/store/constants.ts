export const FETCHED_CART: 'FETCHED_CART' = 'FETCHED_CART'
export const FETCHING_CART: 'FETCHING_CART' = 'FETCHING_CART'
export const ADDED_TO_CART: 'ADDED_TO_CART' = 'ADDED_TO_CART'
export const ADDING_TO_CART: 'ADDING_TO_CART' = 'ADDING_TO_CART'
export const FETCHED_PRODUCT: 'FETCHED_PRODUCT' = 'FETCHED_PRODUCT'
export const FETCHING_PRODUCT: 'FETCHING_PRODUCT' = 'FETCHING_PRODUCT'
export const TOGGLE_CART_OPEN: 'TOGGLE_CART_OPEN' = 'TOGGLE_CART_OPEN'
export const FETCHED_PRODUCTS: 'FETCHED_PRODUCTS' = 'FETCHED_PRODUCTS'
export const FETCHING_PRODUCTS: 'FETCHING_PRODUCTS' = 'FETCHING_PRODUCTS'
export const REMOVING_FROM_CART: 'REMOVING_FROM_CART' = 'REMOVING_FROM_CART'
