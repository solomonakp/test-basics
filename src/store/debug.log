[0216/125848.548:ERROR:registration_protocol_win.cc(102)] CreateFile: The system cannot find the file specified. (0x2)
[0216/150527.038:ERROR:registration_protocol_win.cc(102)] CreateFile: The system cannot find the file specified. (0x2)
[0218/100303.501:ERROR:registration_protocol_win.cc(102)] CreateFile: The system cannot find the file specified. (0x2)
[0218/100318.901:ERROR:registration_protocol_win.cc(102)] CreateFile: The system cannot find the file specified. (0x2)
