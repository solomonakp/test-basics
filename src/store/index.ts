import {
  createStore as createReduxStore,
  combineReducers,
  applyMiddleware,
  compose,
} from 'redux'

import { StoreState } from '../types/Store'
import thunk, { ThunkMiddleware } from 'redux-thunk'
import cart, { cartInitialState } from '../store/reducers/cart'
import products, { productsInitialState } from '../store/reducers/products'

export const initialState: StoreState = {
  products: productsInitialState,
  cart: cartInitialState,
}

export const reducer = combineReducers<StoreState>({
  products,
  cart,
})

const middleware = [
  applyMiddleware(thunk as ThunkMiddleware<StoreState>),
  (window as any).__REDUX_DEVTOOLS_EXTENSION__ &&
        (window as any).__REDUX_DEVTOOLS_EXTENSION__()
].filter(Boolean)

export const createStore = (
  storeReducer = reducer,
  storeInitialState = initialState,
) =>
  createReduxStore(
    storeReducer,
    storeInitialState,
    compose(...middleware),
  )

export default createStore(reducer, initialState)
