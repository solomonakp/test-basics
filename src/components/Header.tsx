import React from 'react'
import styled from 'styled-components'
import { useLocation } from 'react-router-dom'

// hooks
import { useFilters } from '../hooks/useFilters'
import { useDispatch, useSelector } from 'react-redux'
import { toggleCartOpen } from '../store/action-creators'
import { StoreState } from '../types/Store'

const Header = () => {
  const { toggleShowingFilters } = useFilters()
  const { pathname } = useLocation()

  const dispatch = useDispatch()

  const { products } = useSelector<StoreState, StoreState['cart']>(
    (state) => state.cart,
  )

  return (
    <Wrapper>
      <SpanHiddenOnMobile />
      <PageTitle>
        TESTING <span />
        REACT.DEV
      </PageTitle>

      <div>
        {pathname === '/' ? (
          <FilterButton
            data-testid="FilterButton"
            onClick={toggleShowingFilters}
          >
            Filter
          </FilterButton>
        ) : null}
        <CartButton data-testid='CartButton' onClick={() => dispatch(toggleCartOpen())}>
          Cart ({products.length})
        </CartButton>
      </div>
    </Wrapper>
  )
}

const PageTitle = styled.div`
  display: flex;
  align-items: center;
  font-weight: bolder;
`

export const FilterButton = styled.button`
  position: relative;
  height: 33px;
  font-size: 12px;
  font-weight: 500;
  padding: 0px 25px;
  border-radius: 999px;
  border-width: 1px;
  background-color: white;
  cursor: pointer;
  border-style: solid;
  border-color: rgb(229, 229, 229);
  border-image: initial;
  transition: all 0.2s ease-in-out;
  &:hover {
    background-color: rgb(245, 245, 245);
  }
`

const CartButton = styled(FilterButton)`
  margin-left: 10px;
`

const SpanHiddenOnMobile = styled.span`
  display: none;
  @media (min-width: 1024px) {
    display: inline;
  }
`

const Wrapper = styled.header`
  display: flex;
  padding: 10px 20px;
  justify-content: space-between;
`

export default Header
