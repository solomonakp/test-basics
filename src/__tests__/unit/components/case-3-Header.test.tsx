import React from 'react'
import { render, fireEvent } from '@testing-library/react'
import { Provider as StoreProvider } from 'react-redux'
import { createStore } from '../../../store'

import Header from '../../../components/Header'
import { FiltersContext } from '../../../context/filters'
import { MemoryRouter } from 'react-router-dom'

describe('The Header component', () => {
  const defaultContext = {
    toggleShowingFilters: jest.fn(),
  }

  const setupHeader = (value = defaultContext) =>
    render(
      <FiltersContext.Provider value={...value as any}>
        <StoreProvider store={createStore()}>
          <MemoryRouter>
            <Header />
          </MemoryRouter>
        </StoreProvider>
      </FiltersContext.Provider>,
    )
  it('renders header correctly', () => {
    const { asFragment } = setupHeader()

    expect(asFragment()).toMatchSnapshot()
  })

  it('toggles the filter open when the Filter button is clicked', () => {
    const { getByText } = setupHeader()

    const filterButton = getByText(/filter/i)

    fireEvent.click(filterButton)

    expect(defaultContext.toggleShowingFilters).toHaveBeenCalled()
  })

  it('❌shows the filter button only on the home page', () => {})
})
