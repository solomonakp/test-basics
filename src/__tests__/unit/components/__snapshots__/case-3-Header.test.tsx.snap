// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Header component renders header correctly 1`] = `
<DocumentFragment>
  <header
    class="sc-AxhUy ciJHJw"
  >
    <span
      class="sc-AxhCb iSTeBQ"
    />
    <div
      class="sc-AxjAm dGBLMY"
    >
      TESTING 
      <span />
      REACT.DEV
    </div>
    <div>
      <button
        class="sc-AxirZ xKohc"
        data-testid="FilterButton"
      >
        Filter
      </button>
      <button
        class="sc-AxirZ sc-AxiKw lesrjN"
        data-testid="CartButton"
      >
        Cart (0)
      </button>
    </div>
  </header>
</DocumentFragment>
`;
